cmd/oras/login.go

```diff
@@ -74,100 +63,80 @@ Example - Login with insecure registry from command line:
-	cmd.Flags().BoolVarP(&opts.verbose, "verbose", "v", false, "verbose output")
+	option.ApplyFlags(&opts, cmd.Flags())
 	return cmd
 }
 
-func runLogin(opts loginOptions) (err error) {
-	var logLevel logrus.Level
-	if opts.debug {
-		logLevel = logrus.DebugLevel
-	} else if opts.verbose {
-		logLevel = logrus.InfoLevel
-	} else {
-		logLevel = logrus.WarnLevel
-	}
-	ctx, _ := trace.WithLoggerLevel(context.Background(), logLevel)
-
-	// Prepare auth client
-	store, err := credential.NewStore(opts.configs...)
-	if err != nil {
+func preRunLogin(opts loginOptions) (err error) {
+	if err := opts.Credential.ReadPassword(); err != nil {
 		return err
 	}
-
-	// Prompt credential
-	if opts.fromStdin {
-		password, err := io.ReadAll(os.Stdin)
-		if err != nil {
-			return err
-		}
-		opts.password = strings.TrimSuffix(string(password), "\n")
-		opts.password = strings.TrimSuffix(opts.password, "\r")
-	} else if opts.password == "" {
-		if opts.username == "" {
+	if opts.Password == "" {
+		if opts.Username == "" {
+			// prompt for username
 			username, err := readLine("Username: ", false)
 			if err != nil {
 				return err
 			}
-			opts.username = strings.TrimSpace(username)
+			opts.Username = strings.TrimSpace(username)
 		}
-		if opts.username == "" {
-			if opts.password, err = readLine("Token: ", true); err != nil {
+		if opts.Username == "" {
+			// prompt for token
+			if opts.Password, err = readLine("Token: ", true); err != nil {
 				return err
-			} else if opts.password == "" {
+			} else if opts.Password == "" {
 				return errors.New("token required")
 			}
 		} else {
-			if opts.password, err = readLine("Password: ", true); err != nil {
+			// prompt for password
+			if opts.Password, err = readLine("Password: ", true); err != nil {
 				return err
-			} else if opts.password == "" {
+			} else if opts.Password == "" {
 				return errors.New("password required")
 			}
 		}
-	} else {
-		fmt.Fprintln(os.Stderr, "WARNING! Using --password via the CLI is insecure. Use --password-stdin.")
 	}
+	return nil
+}
 
+func runLogin(opts loginOptions) (err error) {
+	ctx, _ := opts.SetLoggerLevel()
+	// Prepare auth client
+	store, err := credential.NewStore(opts.Configs...)
+	if err != nil {
+		return err
+	}
 	// Ping to ensure credential is valid
-	remote, err := remote.NewRegistry(opts.hostname)
+	remote, err := remote.NewRegistry(opts.Hostname)
 	if err != nil {
 		return err
 	}
-	remote.PlainHTTP = opts.plainHTTP
-	cred := credential.Credential(opts.username, opts.password)
-	rootCAs, err := http.LoadCertPool(opts.caFilePath)
+	remote.PlainHTTP = opts.PlainHTTP
+	cred := credential.Credential(opts.Username, opts.Password)
+	config, err := opts.TLS.Config()
 	if err != nil {
 		return err
 	}
 	remote.Client = http.NewClient(http.ClientOptions{
-		Credential:    cred,
-		SkipTLSVerify: opts.insecure,
-		Debug:         opts.debug,
-		RootCAs:       rootCAs,
+		Credential: cred,
+		TLSConfig:  config,
+		Debug:      opts.Debug,
 	})
```

Comment:
This part can also be refactored.

Reply:
It seems that we can combine `option.Credential` with `option.TLS` to be `option.Remote`.

Reply:
The objective is to refactor the code so that can be reused by both registry and repository.

Reply:
Refactored